src/twisted/internet/test/test_inlinecb.py

```diff
@@ -145,3 +147,181 @@ def calling():
         d = calling()
         f = self.failureResultOf(d)
         self.assertIn("in erroring", f.getTraceback())
+
+
+
+class TranslateMe(Exception):
+    """
+    Sample exception type.
+    """
+
+
+
+class TranslateResult(Exception):
+    """
+    Sample exception type.
+    """
+
+
+
+class DontFail(Exception):
+    """
+    Sample exception type.
+    """
+
+    def __init__(self, actual):
+        Exception.__init__(self)
+        self.actualValue = actual
+
+
+
+class CancellationTests(TestCase):
+    """
+    Tests for cancellation of L{Deferred}s returned by L{inlineCallbacks}.
+    For each of these tests, let:
+        - G be a generator decorated with C{inlineCallbacks}
+        - D be a L{Deferred} returned by G
+        - C be a L{Deferred} awaited by G with C{yield}
+    """
+
+    def setUp(self):
+        """
+        Set up the list of outstanding L{Deferred}s.
+        """
+        self.thingsOutstanding = []
+
+
+    def tearDown(self):
+        """
+        If any L{Deferred}s are still outstanding, fire them.
+        """
+        while self.thingsOutstanding:
+            self.thingGotten()
+
+
+    @inlineCallbacks
+    def sampleInlineCB(self, getChildThing=None):
+        """
+        Generator for testing cascade cancelling cases.
+
+        @param getChildThing: Some callable returning L{Deferred} that we
+            awaiting (with C{yield})
+        """
+        if getChildThing is None:
+            getChildThing = self.getThing
+        try:
+            x = yield getChildThing()
+        except TranslateMe:
+            raise TranslateResult()
+        except DontFail as df:
+            x = df.actualValue - 2
+        returnValue(x + 1)
+
+
+    def getThing(self):
+        """
+        A sample function that returns a L{Deferred} that can be fired on
+        demand, by L{CancellationTests.thingGotten}.
+
+        @return: L{Deferred} that can be fired on demand.
+        """
+        self.thingsOutstanding.append(Deferred())
+        return self.thingsOutstanding[-1]
+
+
+    def thingGotten(self, result=None):
+        """
+        Fire the L{Deferred} returned from the least-recent call to
+        L{CancellationTests.getThing}.
+
+        @param result: result object to be used when firing the L{Deferred}.
+        """
+        self.thingsOutstanding.pop(0).callback(result)
+
+
+    def test_cascadeCancellingOnCancel(self):
+        """
+        When D cancelled, C will be immediately cancelled too.
+        """
+        childResultHolder = ['FAILURE']
+        def getChildThing():
+            d = Deferred()
+            def _eb(result):
+                childResultHolder[0] = result.check(CancelledError)
+                return result
+            d.addErrback(_eb)
+            return d
+        d = self.sampleInlineCB(getChildThing=getChildThing)
+        d.addErrback(lambda result: None)
+        d.cancel()
+        self.assertEqual(
```

Comment:
You could use `failureResultOf` instead of `childResultHolder`:

```python
child = defer.Deferred()
d = self.sampleInlineCB(getChildDeferred=lambda: child)
d.cancel()
self.assertRaises(CancelledError, self.failureResultOf(child).raiseException)
```